File: Core/Models/Component.cs
using Core.Enums;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;

namespace Core.Models
{
    public partial class Component
    {
        public string Id { get; set; }
        public string TenantCode { get; set; }
        public string FieldName { get; set; }
        public int? Order { get; set; }
        public string ComponentType { get; set; }
        public string ComponentGroupId { get; set; }
        public string ReferenceId { get; set; }
        public string FormatData { get; set; }
        public string FormatEntity { get; set; }
        public string PlainText { get; set; }
        public int? Column { get; set; }
        public int? Offset { get; set; }
        public int? Row { get; set; }
        public bool CanSearch { get; set; }
        public bool CanCache { get; set; }
        public int? Precision { get; set; }
        public string GroupBy { get; set; }
        public string GroupFormat { get; set; }
        public string Label { get; set; }
        public bool ShowLabel { get; set; }
        public string Icon { get; set; }
        public string ClassName { get; set; }
        public string Style { get; set; }
        public string ChildStyle { get; set; }
        public string HotKey { get; set; }
        public string RefClass { get; set; }
        public string Events { get; set; }
        public bool Disabled { get; set; }
        public bool Visibility { get; set; }
        public bool Hidden { get; set; }
        public string Validation { get; set; }
        public bool Focus { get; set; }
        public string Width { get; set; }
        public string PopulateField { get; set; }
        public string CascadeField { get; set; }
        public string GroupEvent { get; set; }
        public int? XsCol { get; set; }
        public int? SmCol { get; set; }
        public int? LgCol { get; set; }
        public int? XlCol { get; set; }
        public int? XxlCol { get; set; }
        public string DefaultVal { get; set; }
        public string DateTimeField { get; set; }
        public bool Active { get; set; }
        public DateTimeOffset InsertedDate { get; set; }
        public string InsertedBy { get; set; }
        public DateTimeOffset? UpdatedDate { get; set; }
        public string UpdatedBy { get; set; }
        public string RoleId { get; set; }
        public bool IgnoreSync { get; set; }
        public bool CanAdd { get; set; }
        public bool ShowAudit { get; set; }
        public bool IsPrivate { get; set; }
        public string IdField { get; set; }
        public string DescFieldName { get; set; }
        public int? MonthCount { get; set; }
        public bool? IsDoubleLine { get; set; }
        public string Query { get; set; }
        public string LocalQuery { get; set; }
        public bool IsRealtime { get; set; }
        public string RefName { get; set; }
        public bool TopEmpty { get; set; }
        public bool IsCollapsible { get; set; }
        public string Template { get; set; }
        public string Renderer { get; set; }
        public string PreQuery { get; set; }
        public string DisabledExp { get; set; }
        public bool FocusSearch { get; set; }
        public bool IsSumary { get; set; }
        public string FormatSumaryField { get; set; }
        public string OrderBySumary { get; set; }
        public bool ShowHotKey { get; set; }
        public int? DefaultAddStart { get; set; }
        public int? DefaultAddEnd { get; set; }
        public bool UpperCase { get; set; }
        public bool VirtualScroll { get; set; }
        public string Migration { get; set; }
        public string ListClass { get; set; }
        public string ExcelFieldName { get; set; }
        public bool LiteGrid { get; set; }
        public bool ShowDatetimeField { get; set; }
        public bool ShowNull { get; set; }
        public bool AddDate { get; set; }
        public bool FilterEq { get; set; }
        public int? HeaderHeight { get; set; }
        public int? BodyItemHeight { get; set; }
        public int? FooterHeight { get; set; }
        public int? ScrollHeight { get; set; }
        public string ScriptValidation { get; set; }
        public bool FilterLocal { get; set; }
        public bool HideGrid { get; set; }
        public string GroupReferenceId { get; set; }
        public string GroupReferenceName { get; set; }
        public string GroupName { get; set; }
        public string ShortDesc { get; set; }
        public string Description { get; set; }
        public string FeatureId { get; set; }
        public string EntityId { get; set; }
        public string EntityName { get; set; }
        public string ComponentId { get; set; }
        public string TextAlign { get; set; }
        public bool HasFilter { get; set; }
        public bool Frozen { get; set; }
        public string FilterTemplate { get; set; }
        public bool Editable { get; set; }
        public string FormatExcell { get; set; }
        public string DatabaseName { get; set; }
        public string Summary { get; set; }
        public int? SummaryColSpan { get; set; }
        public bool BasicSearch { get; set; }
        public bool IsExport { get; set; }
        public int? OrderExport { get; set; }
        public string ShowExp { get; set; }
        public string MinWidth { get; set; }
        public string MaxWidth { get; set; }
        public bool AdvancedSearch { get; set; }
        public bool AutoFit { get; set; }
        public bool DisplayNone { get; set; }
        public string FieldText { get; set; }
        public string OrderBy { get; set; }
        public string QueueName { get; set; }
        public string ConnKey { get; set; }
        public bool ShouldSaveText { get; set; }
        public string CacheName { get; set; }

        [DbIgnore]
        public TextAlign? TextAlignEnum { get; set; }
        [DbIgnore]
        public bool IsPivot { get; set; }
        [DbIgnore]
        public int PostOrder { get; set; }
        [DbIgnore]
        public List<object> LocalData { get; set; }
        [DbIgnore]
        public List<Component> LocalHeader { get; set; }
        [DbIgnore]
        public bool StatusBar { get; set; }
        [DbIgnore]
        public bool SimpleText { get; set; }
        [DbIgnore]
        public bool LocalRender { get; set; }
        [DbIgnore]
        public bool IgnoreConfirmHardDelete { get; set; }

        [DbIgnore]
        public virtual ComponentGroup ComponentGroup { get; set; }
        [DbIgnore]
        public virtual Entity Reference { get; set; }
    }
}

File: CoreAPI/wwwRoot/js/core/models/component.js
/** @typedef {import('./validationRule.js').ValidationRule} ValidationRule */
/**
 * Represents a Component.
 * @class Component
 * @property {string} Id
 * @property {string} TenantCode
 * @property {string} FieldName
 * @property {number|null} Order
 * @property {string} ComponentType
 * @property {string} ComponentGroupId
 * @property {string} ReferenceId
 * @property {string} FormatData
 * @property {string} FormatEntity
 * @property {string} PlainText
 * @property {number|null} Column
 * @property {number|null} Offset
 * @property {number|null} Row
 * @property {boolean} CanSearch
 * @property {boolean} CanCache
 * @property {number|null} Precision
 * @property {string} GroupBy
 * @property {string} GroupFormat
 * @property {string} Label
 * @property {boolean} ShowLabel
 * @property {string} Icon
 * @property {string} ClassName
 * @property {string} Style
 * @property {string} ChildStyle
 * @property {string} HotKey
 * @property {string} RefClass
 * @property {string} Events
 * @property {boolean} Disabled
 * @property {boolean} Visibility
 * @property {boolean} Hidden
 * @property {string | ValidationRule[]} Validation
 * @property {boolean} Focus
 * @property {string} Width
 * @property {string} PopulateField
 * @property {string} CascadeField
 * @property {string} GroupEvent
 * @property {number|null} XsCol
 * @property {number|null} SmCol
 * @property {number|null} LgCol
 * @property {number|null} XlCol
 * @property {number|null} XxlCol
 * @property {string} DefaultVal
 * @property {string} DateTimeField
 * @property {boolean} Active
 * @property {Date} InsertedDate
 * @property {string} InsertedBy
 * @property {Date|null} UpdatedDate
 * @property {string} UpdatedBy
 * @property {string} RoleId
 * @property {boolean} IgnoreSync
 * @property {boolean} CanAdd
 * @property {boolean} ShowAudit
 * @property {boolean} IsPrivate
 * @property {string} IdField
 * @property {string} DescFieldName
 * @property {number|null} MonthCount
 * @property {boolean|null} IsDoubleLine
 * @property {string} Query
 * @property {string} LocalQuery
 * @property {boolean} IsRealtime
 * @property {string} RefName
 * @property {boolean} TopEmpty
 * @property {boolean} IsCollapsible
 * @property {string} Template
 * @property {string} Renderer
 * @property {string} PreQuery
 * @property {string} DisabledExp
 * @property {boolean} FocusSearch
 * @property {boolean} IsSumary
 * @property {string} FormatSumaryField
 * @property {string} OrderBySumary
 * @property {boolean} ShowHotKey
 * @property {number|null} DefaultAddStart
 * @property {number|null} DefaultAddEnd
 * @property {boolean} UpperCase
 * @property {boolean} VirtualScroll
 * @property {string} Migration
 * @property {string} ListClass
 * @property {string} ExcelFieldName
 * @property {boolean} LiteGrid
 * @property {boolean} ShowDatetimeField
 * @property {boolean} ShowNull
 * @property {boolean} AddDate
 * @property {boolean} FilterEq
 * @property {number|null} HeaderHeight
 * @property {number|null} BodyItemHeight
 * @property {number|null} FooterHeight
 * @property {number|null} ScrollHeight
 * @property {string} ScriptValidation
 * @property {boolean} FilterLocal
 * @property {boolean} HideGrid
 * @property {string} GroupReferenceId
 * @property {string} GroupReferenceName
 * @property {string} GroupName
 * @property {string} Description
 * @property {string} FeatureId
 * @property {string} EntityId
 * @property {string} EntityName
 * @property {string} ComponentId
 * @property {string} TextAlign
 * @property {boolean} HasFilter
 * @property {boolean} Frozen
 * @property {string} FilterTemplate
 * @property {boolean} Editable
 * @property {string} FormatExcell
 * @property {string} DatabaseName
 * @property {string} Summary
 * @property {number|null} SummaryColSpan
 * @property {boolean} IsExport
 * @property {number|null} OrderExport
 * @property {string} ShowExp
 * @property {string} MinWidth
 * @property {string} MaxWidth
 * @property {boolean} AdvancedSearch
 * @property {boolean} AutoFit
 * @property {boolean} DisplayNone
 * @property {string} FieldText
 * @property {string} OrderBy
 * @property {string} QueueName
 * @property {string} MetaConn
 * @property {string} DataConn
 * @property {boolean} ShouldSaveText
 * @property {string} CacheName
 * @property {string} Lang
 * @property {string} DelCmd
 * @property {string} DelParam
 */

import { FeaturePolicy } from './featurePolicy.js';

// Initializing data
export class Component {
    Id = '';
    TenantCode = '';
    Name = '';
    FieldName = '';
    Order = 0;
    ComponentType = '';
    ComponentGroupId = '';
    ReferenceId = '';
    FormatData = '';
    FormatEntity = '';
    PlainText = '';
    Column = null;
    Offset = null;
    Row = null;
    CanSearch = false;
    CanCache = false;
    Precision = null;
    GroupBy = '';
    GroupFormat = '';
    Label = '';
    ShowLabel = false;
    Icon = '';
    ClassName = '';
    Style = '';
    ChildStyle = '';
    HotKey = '';
    RefClass = '';
    Events = '';
    Disabled = false;
    Visibility = false;
    Hidden = false;
    /** @type {String | ValidationRule[]} */
    Validation;
    Focus = false;
    Width = '';
    PopulateField = '';
    CascadeField = '';
    GroupEvent = '';
    XsCol = null;
    SmCol = null;
    LgCol = null;
    XlCol = null;
    XxlCol = null;
    DefaultVal = '';
    DateTimeField = '';
    Active = false;
    InsertedDate = new Date();
    InsertedBy = '';
    UpdatedDate = null;
    UpdatedBy = '';
    RoleId = '';
    IgnoreSync = false;
    CanAdd = false;
    ShowAudit = false;
    IsPrivate = false;
    IdField = '';
    DescFieldName = '';
    MonthCount = null;
    IsDoubleLine = null;
    Query = '';
    /** @type {String | any[]} */
    LocalQuery = '';
    IsRealtime = false;
    RefName = '';
    TopEmpty = false;
    IsCollapsible = false;
    Template = '';
    /** @type {String | function} */
    Renderer;
    PreQuery = '';
    DisabledExp = '';
    FocusSearch = false;
    IsSumary = false;
    FormatSumaryField = '';
    OrderBySumary = '';
    ShowHotKey = false;
    DefaultAddStart = null;
    DefaultAddEnd = null;
    UpperCase = false;
    VirtualScroll = false;
    Migration = '';
    ListClass = '';
    ExcelFieldName = '';
    LiteGrid = false;
    ShowDatetimeField = false;
    ShowNull = false;
    AddDate = false;
    FilterEq = false;
    HeaderHeight = null;
    BodyItemHeight = null;
    FooterHeight = null;
    ScrollHeight = null;
    ScriptValidation = '';
    FilterLocal = false;
    HideGrid = false;
    GroupReferenceId = '';
    GroupReferenceName = '';
    GroupName = '';
    Description = '';
    FeatureId = '';
    EntityId = '';
    EntityName = '';
    ComponentId = '';
    TextAlign = '';
    HasFilter = false;
    Frozen = false;
    FilterTemplate = '';
    Editable = false;
    FormatExcell = '';
    DatabaseName = '';
    Summary = '';
    SummaryColSpan = null;
    IsExport = false;
    OrderExport = null;
    ShowExp = '';
    MinWidth = '';
    MaxWidth = '';
    AdvancedSearch = false;
    AutoFit = false;
    DisplayNone = false;
    FieldText = '';
    OrderBy = '';
    QueueName = '';
    MetaConn = '';
    DataConn = '';
    ShouldSaveText = false;
    CacheName = '';
    Lang = '';
    DelCmd = '';
    DelParam = '';
    DisplayField = '';
    DisplayDetail = '';
    TextAlignEnum = null;
    IsPivot = false;
    PostOrder = 0;
    LocalData = [];
    /** @type {Component[]} */
    LocalHeader = [];
    StatusBar = false;
    SimpleText = false;
    LocalRender = false;
    IgnoreConfirmHardDelete = false;
    ComponentGroup = null;
    Reference = null;
    /**
     * Represents whether the component is a tab.
     * @type {boolean}
     */
    IsTab;

    /**
     * Represents the group to which the tab belongs.
     * @type {string}
     */
    TabGroup;

    /**
     * Represents whether the tab is vertical.
     * @type {boolean}
     */
    IsVerticalTab;

    /**
     * Represents whether the component is responsive.
     * @type {boolean}
     */
    Responsive;

    /**
     * Represents the policy ID associated with the component.
     * @type {string}
     */
    PolicyId;

    /**
     * Represents the parent of the component.
     * @type {Component}
     */
    Parent;

    /**
     * Represents the parent of the component.
     * @type {Component[]}
     */
    Component = [];

    /**
     * Represents the parent ID of the component.
     * @type {string}
     */
    ParentId;

    /**
     * Represents the number of outer columns.
     * @type {number | null}
     */
    OuterColumn;

    /**
     * Represents the number of outer columns for extra small screens.
     * @type {number | null}
     */
    XsOuterColumn;

    /**
     * Represents the number of outer columns for small screens.
     * @type {number | null}
     */
    SmOuterColumn;

    /**
     * Represents the number of outer columns for large screens.
     * @type {number | null}
     */
    LgOuterColumn;

    /**
     * Represents the number of outer columns for extra large screens.
     * @type {number | null}
     */
    XlOuterColumn;

    /**
     * Represents the number of outer columns for extra extra large screens.
     * @type {number | null}
     */
    XxlOuterColumn;

    /**
     * Represents the month for which a badge is displayed.
     * @type {number | null}
     */
    BadgeMonth;

    /**
     * Represents whether the component is a drop-down.
     * @type {boolean}
     */
    IsDropDown;

    /**
     * Represents the number of items in a row.
     * @type {number | null}
     */
    ItemInRow;

    /**
     * Represents the HTML content of the component.
     * @type {string}
     */
    Html;

    /**
     * Represents the CSS styles of the component.
     * @type {string}
     */
    Css;

    /**
     * Represents the JavaScript code associated with the component.
     * @type {string}
     */
    Javascript;
    
    /**
     * If set to true, then not to encode the entire children
     * @type {Boolean}
     */
    IgnoreEncode;
    /** @type {FeaturePolicy[]} */
    FeaturePolicy = [];

    /** @type {Component[]} */
    Children = [];
};